src/ir/iteration.h

```diff
@@ -57,8 +70,11 @@ template<class Specific> class AbstractChildIterator {
 
     void operator++() { index++; }
 
-    Expression*& operator*() {
```

Comment:
I'm not sure I understand the motivation here yet. The old code returns an `Expression*&` which also allows replacing the child, the same as returning `**` - the reference is to the actual pointer in the parent expression. This works as the `children` vector contains `Expression**` already (which I see this PR does not change).

Is the goal here to return `**` instead for clarity reasons? Or something else?

Reply:
I tried to do this using iterator assignment based on the existing iterator but it was hard. So what I'm doing is this: https://github.com/aheejin/binaryen/blob/844cda5efcafda97c796bf20bb926d6a14749abb/src/ir/eh-utils.cpp#L26-L90

So what I've been trying to do is make a function that returns a pointer (`Expression**`) to a nested `pop` and use that pointer to assign another expression (`local.get`) in that place. The function's signature is like
```cpp
Expression* getFirstPop(Expresion* catchBody, bool& isPopNested, Expression**& popPtr);
```
So this returns the pointer to `pop`, and also returns a pointer to it as an output parameter. I don't particularly like this design though. The reason I made it return both of them is, if the catch's body itself is a `pop`, it is possible that we can get `Expression*` but not `Expression**`

I tried to make this function to return `ChildIterator::Iterator` at your suggestion, but it was somehow not easy. First of all, `ChildIterator::Iterator` is a private type so we have to make it public to use it as a function return type. Maybe that's fine. But the current structure of the `while` loop in that function is, you create an iterator at the bottom of the loop and return something at the top of the loop; the reason it is structured in this way is to handle the case that the given catch body itself is a `pop`. To make an iterator, we need a way for an iterator variable to hold its values across loop iterations, and because `ChildIterator::Iterator` doesn't have a default constructor, it was hard... Maybe I'm missing something and there's an easier way.


I guess I think I can return an iterator instead, so that I can use it to both dereference it and assign something else to its place...? I'll try that.

Reply:
I think I see, thanks.

For getting the `**`, I think in C++ it is possible to convert a reference to a pointer, that is,
```cpp
void foo(T*& t) {
  T** p = &t; // convert &* into **
}
```
It's possible that's undefined behavior, though..? Not sure. But at the machine level those are the same... And in practice this is how we replace children today using this API, effectively.

For `getFirstPop`, maybe instead of returning `**` as an out param you can return a `std::variant` that has one option be a `*&`? Not sure if that's possible though, or if it's a good idea if it is...

Reply:
Oh, yeah, I can do that... Why was I thinking that was not possible.. 😨 Thanks, and sorry @tlively for taking your time... I guess than we can close this?

Not sure `std::variant` would be good for that purpose... Anyway we can talk more on that PR.